### PR TITLE
Simple early termination iterator

### PR DESCRIPTION
**Pull request by eraly**
## What changes were proposed in this pull request?
https://github.com/deeplearning4j/deeplearning4j/issues/3442

## How was this patch tested?
Unit tests

### CODE REVIEW

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationMultiDataSetIterator.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+     * @param terminationPoint, minibatches after which hasNext() will return false
+     */
+    public EarlyTerminationMultiDataSetIterator(MultiDataSetIterator underlyingIterator, int terminationPoint) {
+        if (terminationPoint <= 0 ) throw new IllegalArgumentException("Termination point (the number of calls to .next() or .next(num)) has to be > 0");
+        this.underlyingIterator = underlyingIterator;
+        this.terminationPoint = terminationPoint;
+    }
+
+    @Override
+    public MultiDataSet next(int num) {
```
Comment (huitseeker) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationMultiDataSetIterator.java]: if minibatchCount > (terminationPoint - num), before `next(num)` this call with succeed, returning a total # of elements strictly greater than terminationPoint

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationDataSetIterator.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+     */
+    public EarlyTerminationDataSetIterator(DataSetIterator underlyingIterator, int terminationPoint) {
+        if (terminationPoint <= 0 ) throw new IllegalArgumentException("Termination point (the number of calls to .next() or .next(num)) has to be > 0");
+        this.underlyingIterator = underlyingIterator;
+        this.terminationPoint = terminationPoint;
+    }
+
+    @Override
+    public DataSet next(int num) {
+        if (minibatchCount < terminationPoint) {
```
Comment (huitseeker) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationDataSetIterator.java]: if minibatchCount > (terminationPoint - num), before next(num) this call with succeed, returning a total # of elements strictly greater than terminationPoint

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationDataSetIterator.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        return underlyingIterator.getLabels();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return underlyingIterator.hasNext() && minibatchCount < terminationPoint;
+    }
+
+    @Override
+    public DataSet next() {
```
Comment (huitseeker) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationDataSetIterator.java]: This call (and the associated one with `next(num)` does not seem to be thread-safe. Should we sync on the miniBatchCount, or make it Atomic ? Or is it better to deal with this by adding javadoc on how this class is not thread-safe ?

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationMultiDataSetIterator.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        underlyingIterator.reset();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return underlyingIterator.hasNext() && minibatchCount < terminationPoint;
+    }
+
+    @Override
+    public MultiDataSet next() {
```
Comment (huitseeker) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationMultiDataSetIterator.java]: This call (and the associated one with `next(num)` does not seem to be thread-safe. Should we sync on the miniBatchCount, or make it Atomic ? Or is it better to deal with this by adding javadoc on how this class is not thread-safe ?

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationDataSetIterator.java (hunk 3 above)
Comment (raver119) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationDataSetIterator.java]: All our iterator implementations are not thread safe, and they are never used in concurrent way.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationDataSetIterator.java (hunk 2 above)
Comment (eraly) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationDataSetIterator.java]: Termination point is the number of minibatches.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationMultiDataSetIterator.java (hunk 1 above)
Comment (eraly) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationMultiDataSetIterator.java]: The termination point is not the number of elements. It is the number of minibatches. I have made that clear in the javadoc. 
https://github.com/deeplearning4j/deeplearning4j/pull/3502/files#diff-b986a13d8f51bf0dae2466fc7faa4d52R11

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationMultiDataSetIterator.java (hunk 4 above)
Comment (huitseeker) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationMultiDataSetIterator.java]: Answered by raver

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationDataSetIterator.java (hunk 2 above)
Comment (huitseeker) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationDataSetIterator.java]: Ah, ok, sory for the overlook. In that case, how about changing the error message to mention the "budgeted # of minibatches"  explicitly ?

Comment (eraly) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/EarlyTerminationDataSetIterator.java]: Done.

**Comment by eraly** [comment]
Thanks for the review @huitseeker Merging.

**Pull request merged**